File: Python/Test_gFN_Computation.py
# Yuncong Ma, 9/13/2023
# Test the computation of gFNs

#########################################
# Packages
import nibabel as nib
import numpy as np
import scipy.io as sio
import os
import re
from scipy.io import savemat
from scipy.sparse import spdiags, lil_matrix
from scipy.spatial import distance
import time

import pNet


os.system('export OMP_NUM_THREADS=1')
os.system('export OPENBLAS_NUM_THREADS=1')

#########################################
# Example data from HCP surf Test_FN17
# gNb uses index starting from 1, but it starts from 0 in Python
gNb = pNet.load_matlab_single_array(os.path.join(pNet.Example.HCP_surf.dir_pnet, 'FN_Computation', 'gNb.mat'))
Data = pNet.load_matlab_single_array(os.path.join(pNet.Example.HCP_surf.dir_data, '100206', '1', 'LR', 'Image.mat'))
Data = Data.T
Setting = pNet.load_json_setting(os.path.join(pNet.Example.HCP_surf.dir_pnet, 'FN_Computation', 'Setting.json'))

# Setup parameters
K = 17
maxIter = 1000
minIter = 30
error = 1e-6
normW = 1
eta = 0
Alpha = 2
Beta = 30
vxI = 0
ard = 0
dataPrecision = 'double'
logFile = 'Log_gFN_NMF.log'

# Internal parameters
alphaS = 0
alphaL = 0

# setup log file
logFile = open(logFile, 'w')
print(f'\nStart NMF for gFN using NumPy at '+time.strftime('%Y-%m-%d %H:%M:%S', time.localtime(time.time()))+'\n', file=logFile)

# Input data size
dim_time, dim_space = Data.shape

# Median number of graph neighbors
nM = np.median(np.unique(gNb[:, 0], return_counts=True)[1])

# Use Alpha and Beta to set alphaS and alphaL if they are 0
if alphaS == 0 and Alpha > 0:
    alphaS = np.round(Alpha * dim_time / K)
if alphaL == 0 and Beta > 0:
    alphaL = np.round(Beta * dim_time / K / nM)

# Setup data precision and eps
np_float, np_eps = pNet.set_data_precision(dataPrecision)

Data = Data.astype(np_float)

# Prepare and normalize scan
X = pNet.normalize_data(Data, 'vp', 'vmax')

# Initialize U and V
mean_X = np.sum(X) / (dim_time*dim_space)
rng = np.random.default_rng(1)
U = (rng.random((dim_time, K), dtype=np_float) + 1) * (np.sqrt(mean_X/K))
V = (rng.random((dim_space, K), dtype=np_float) + 1) * (np.sqrt(mean_X/K))

# To ensure the same results
U = pNet.load_matlab_single_array('gFN_init_U.mat')
V = pNet.load_matlab_single_array('gFN_init_V.mat')

# Normalize data
U, V = pNet.normalize_u_v(U, V, 1, 1)

# Construct the spatial affinity graph
L, W, D = pNet.construct_Laplacian_gNb(gNb, dim_space, vxI, X, alphaL, normW, dataPrecision)

if ard > 0:
    ard = 1
    eta = 0.1
    lambdas = np.sum(U, axis=0) / dim_time
    hyperLam = eta * np.sum(np.power(X, 2)) / (dim_time * dim_space * 2)
else:
    lambdas = 0
    hyperLam = 0

oldLogL = np.inf

# Multiplicative update of U and V
for i in range(1, 1+maxIter):
    # ===================== update V ========================
    # Eq. 8-11
    XU = X.T @ U
    UU = U.T @ U
    VUU = V @ UU

    tmpl2 = np.power(V, 2)

    if alphaS > 0:
        tmpl21 = np.sqrt(tmpl2)
        tmpl22 = np.tile(np.sqrt(np.sum(tmpl2, axis=0)), (dim_space, 1))
        tmpl21s = np.tile(np.sum(tmpl21, axis=0), (dim_space, 1))
        posTerm = V / np.maximum(tmpl21 * tmpl22, np_eps)
        negTerm = V * tmpl21s / np.maximum(np.power(tmpl22, 3), np_eps)

        VUU = VUU + 0.5 * alphaS * posTerm
        XU = XU + 0.5 * alphaS * negTerm

    if alphaL > 0:
        WV = W @ V.astype(np.float64)
        DV = D @ V.astype(np.float64)

        XU = XU + WV
        VUU = VUU + DV

    V = V * (XU / np.maximum(VUU, np_eps))

    # Prune V if empty components are found in V
    # This is almost impossible to happen without combining FNs
    prunInd = np.sum(V != 0, axis=0) == 1
    if np.any(prunInd):
        V[:, prunInd] = np.zeros((dim_space, np.sum(prunInd)))
        U[:, prunInd] = np.zeros((dim_time, np.sum(prunInd)))

    # normalize U and V
    U, V = pNet.normalize_u_v(U, V, 1, 1)

    # ===================== update U =========================
    XV = X @ V
    VV = V.T @ V
    UVV = U @ VV

    if ard > 0:  # ard term for U
        posTerm = 1 / np.maximum(np.tile(lambdas, (dim_time, 1)), np_eps)

        UVV = UVV + posTerm * hyperLam

    U = U * (XV / np.maximum(UVV, np_eps))

    # Prune U if empty components are found in U
    # This is almost impossible to happen without combining FNs
    prunInd = np.sum(U, axis=0) == 0
    if np.any(prunInd):
        V[:, prunInd] = np.zeros((dim_space, np.sum(prunInd)))
        U[:, prunInd] = np.zeros((dim_time, np.sum(prunInd)))

    # update lambda
    if ard > 0:
        lambdas = np.sum(U) / dim_time

    # ==== calculate objective function value ====
    ardU = 0
    tmp1 = 0
    tmp2 = 0
    tmp3 = 0
    tmpl21 = np.power(V, 2)

    if ard > 0:
        su = np.sum(U, axis=0)
        su[su == 0] = 1
        ardU = np.sum(np.log(su)) * dim_time * hyperLam

    if alphaL > 0:
        tmp2 = V.T @ L * V.T

    L21 = alphaS * np.sum(np.sum(np.sqrt(tmpl21), axis=0) / np.maximum(np.sqrt(np.sum(tmpl21, axis=0)), np_eps))
    # LDf = pNet.data_fitting_error(X, U, V, 0, 1)
    LDf = np.sum(np.power(X - U @ V.T, 2))
    LSl = np.sum(tmp2)

    # Objective function
    LogL = L21 + LDf + LSl + ardU
    print(f"    Iter = {i}: LogL: {LogL}, dataFit: {LDf}, spaLap: {LSl}, L21: {L21}, ardU: {ardU}", file=logFile)

    # The iteration needs to meet minimum iteration number and small changes of LogL
    if i > minIter and abs(oldLogL - LogL) / np.maximum(oldLogL, np_eps) < error:
        break
    oldLogL = LogL.copy()

savemat('gFN_U_end.mat', {'U': U})
savemat('gFN_V_end.mat', {'V': V})

print(f'\n Finished at '+time.strftime('%Y-%m-%d %H:%M:%S', time.localtime(time.time()))+'\n', file=logFile)

File: MATLAB/Private_Function/MATLAB_Yuncong/Development/Test_gFN_Computation.m
% Yuncong Ma, 8/21/2023
% This script is to test the corresponding python file, in order to ensure
% the same result of gFN computation
% Test_gFN_Computation.py

%% Add to path
addpath(genpath('/Users/yuncongma/Documents/Document/fMRI/Myworks/pNet'));
%%
maxNumCompThreads(10);

% load file
Data = fLoad_MATLAB_Single_Variable('/Users/yuncongma/Documents/Document/fMRI/Myworks/pNet/Example/HCP_Surface/Data/100206/1/LR/Image.mat');
gNb = fLoad_MATLAB_Single_Variable('/Users/yuncongma/Documents/Document/fMRI/Myworks/pNet/Example/HCP_Surface/Test_FN17/FN_Computation/gNb.mat');
Setting = fLoad_MATLAB_Single_Variable('/Users/yuncongma/Documents/Document/fMRI/Myworks/pNet/Example/HCP_Surface/Test_FN17/FN_Computation/Setting.mat');

Data=Data';
%%
tic
% Setup parameter
K = 17;
maxIter = 1000;
minIter = 30;
error = 1e-6;
normW = 1;
eta = 0;
Alpha = 2;
Beta = 30;
vxI = 0;
ard = 0;
dataPrecision = 'double';
logFile = 'Log_gFN_NMF_MATLAB.log';

% Internal parameter
alphaS = 0;
alphaL = 0;

% setup log file
% fileID = fopen(logFile, 'w');
fprintf(['\nStart NMF for gFN using MATLAB at ',char(datetime),'\n'])


% [mat_float,mat_eps] = set_data_precision(dataPrecision);
mat_eps = eps(dataPrecision);
if strcmp(dataPrecision,'single')
    Data=single(Data);
else
    Data=double(Data);
end

dim_space = size(Data,2);
dim_time = size(Data,1);

nmVec = zeros(length(gNb),1);
for gni=1:length(gNb)
    nmVec(gni) = length(gNb{gni});
end
nM = median(nmVec);

alphaS = round((Alpha*dim_time)/K);
alphaL = round((Beta*dim_time)/(K*nM));

% normalize data
X = normalize_data(Data,'vp','vmax');

% Initialize U and V
mean_X = sum(X(:))/(dim_time*dim_space);
rng('default')
U = (rand(dim_time,K)+1)*(sqrt(mean_X/K));
V = (rand(dim_space,K)+1)*(sqrt(mean_X/K));

% save('./gFN_init_U.mat','U');
% save('./gFN_init_V.mat','V');

[U,V] = normalize_u_v(U, V, 1, 1);

% construct the spatial affinity graph
[L,W,D] = construct_Laplacian_gNb(gNb, X, vxI, dim_space, alphaL, normW, dataPrecision);


if ard>0
    ard = 1;
    eta = 0.1;
    lambdas = sum(U,1) / dim_time;
    hyperLam = eta * sum(sum(X.^2,1),2) / (dim_time*dim_space*2);
else
    lambdas=0;
    hyperLam=0;
end

oldLogL=inf;

% Multiplicative update of U and V
for i=1:maxIter

    % ===================== update V ========================
    XU = X'*U;  % mnk or pk (p<<mn)
    UU = U'*U;  % mk^2
    VUU = V*UU; % nk^2

    if alphaS>0
        %             % L1 sparsity
        %             F = ones(size(V));
        %             VUU = VUU + 0.5*alphaS * F;

        % scale-invariant sparsity
        tmpNorm2 = sqrt(sum(V.^2,1));
        posTerm = 1 ./ max(repmat(tmpNorm2,dim_space,1),mat_eps);
        tmpNorm1 = sum(V,1);
        negTerm = V .* repmat(tmpNorm1,dim_space,1) ./ max(repmat(tmpNorm2.^3,dim_space,1),mat_eps);

        XU = XU + 0.5*alphaS * negTerm;
        VUU = VUU + 0.5*alphaS * posTerm;
    end

    if alphaL>0
        V = double(V);
        WV = W*V;
        DV = D*V;

        XU = XU + WV;
        VUU = VUU + DV;
    end

    V = V.*(XU./max(VUU,mat_eps));

    prunInd = sum(V~=0)==1;
    if any(prunInd)
        V(:,prunInd) = zeros(dim_space,sum(prunInd));
        U(:,prunInd) = zeros(dim_time,sum(prunInd));
    end

    [U, V] = normalize_u_v(U, V, 1, 1);

    % ===================== update U ========================
    XV = X*V;   % mnk or pk (p<<mn)
    VV = V'*V;  % nk^2
    UVV = U*VV; % mk^2

    if ard>0
        posTerm = 1./max(repmat(lambdas,dim_time,1),mat_eps);
        UVV = UVV + posTerm*hyperLam;
    end

    U = U.*(XV./max(UVV,eps));

    prunInd = sum(U)==0;
    if any(prunInd)
        V(:,prunInd) = zeros(dim_space,sum(prunInd));
        U(:,prunInd) = zeros(dim_time,sum(prunInd));
    end

    if ard>0
        lambdas = sum(U) / dim_time;
    end
    
    % ==== calculate partial objective function value ====
    ardU = 0;
    tmp1 = 0;
    tmp2 = 0;
    tmp3 = 0;
    tmpl21 = V.^2;

    if ard>0
        su=sum(U,1);
        su(su==0)=1;
        ardU=sum(log(su))*dim_time*hyperLam;
    end

    tmpDf=(X-U*V').^2;
    tmp1=sum(tmpDf(:));

    if alphaL>0
        tmp2=V'*L.*V';
    end

    L21=alphaS*sum(sum(sqrt(tmpl21),1)./max(sqrt(sum(tmpl21,1)),mat_eps));
    LDf = tmp1;
    LSl = sum(tmp2(:));

    % Objective function
    LogL = L21 + ardU + LDf + LSl;

    fprintf(['  Iter = ',num2str(i),': LogL:',num2str(LogL),',dataFit:',num2str(LDf)...
        ',spaLap:',num2str(LSl),...
        ',L21:',num2str(L21),',ardU:',num2str(ardU),'\n']);
    
    if i>minIter && abs(oldLogL-LogL)/max(oldLogL,mat_eps)<error
        break;
    end
    oldLogL = LogL;
end


toc
